util/feeding.py

```diff
@@ -1,13 +1,109 @@
+import numpy as np
+import os
 import pandas
+import tables
 import tensorflow as tf
 
-from threading import Thread
+from functools import partial
 from math import ceil
+from multiprocessing.dummy import Pool
 from six.moves import range
+from threading import Thread
 from util.audio import audiofile_to_input_vector
 from util.gpu import get_available_gpus
 from util.text import ctc_label_dense_to_sparse, text_to_char_array
 
+
+def pmap(fun, iterable, threads=8):
+    pool = Pool(threads)
+    results = pool.map(fun, iterable)
+    pool.close()
+    return results
+
+
+def process_single_file(row, numcep, numcontext, alphabet):
+    # row = index, Series
+    _, file = row
+    features = audiofile_to_input_vector(file.wav_filename, numcep, numcontext)
+    transcript = text_to_char_array(file.transcript, alphabet)
+
+    if (2*numcontext + len(features)) < len(transcript):
+        raise ValueError('Error: Audio file {} is too short for transcription.'.format(file.wav_filename))
+
+    return features, len(features), transcript, len(transcript)
+
+
+# load samples from CSV, compute features, optionally cache results on disk
+def preprocess(csv_files, batch_size, numcep, numcontext, alphabet, hdf5_cache_path=None):
```

Comment:
How about putting `preprocess` and its dependencies into an extra file that would have no TF dependencies and so on? Then we could also provide a simple command-line wrapper around it (under `/bin`).

Reply:
OK.